### PR TITLE
DF-936 caching optimization for the occupation label repository

### PR DESCRIPTION
**Pull request by ganqqwerty**

### CODE REVIEW

**Comment by ganqqwerty** [comment]
was versus became:
image

File: alv-portal-ui/src/app/shared/backend-services/reference-service/occupation-label.repository.ts [hunk 1]
```diff
@@ -1,13 +1,11 @@
...
-  OccupationLabelAutocomplete,
-  OccupationLabelData
-} from './occupation-label.types';
-import { tap } from 'rxjs/operators';
+import { Observable } from 'rxjs';
+import { OccupationLabelAutocomplete, OccupationLabelData } from './occupation-label.types';
+import { shareReplay } from 'rxjs/operators';
 
 const DEFAULT_RESPONSE_SIZE = '10';
+const CACHE_SIZE = 1;
```
Comment (99999989) [inline alv-portal-ui/src/app/shared/backend-services/reference-service/occupation-label.repository.ts]: After reading the `shareReplay` docs I would recommend to call it `bufferSize` (as in the shareReplay config).

File: alv-portal-ui/src/app/shared/backend-services/reference-service/occupation-label.repository.ts [hunk 2]
```diff
@@ -20,24 +18,23 @@ export enum OccupationTypes {
   SBN5 = 'SBN5',
 }
 
-@Injectable({ providedIn: 'root' })
+@Injectable({providedIn: 'root'})
```
Comment (99999989) [inline alv-portal-ui/src/app/shared/backend-services/reference-service/occupation-label.repository.ts]: This formatting is wrong afaik. Didn't you import the ALV.xml for code formatting some time ago.

File: alv-portal-ui/src/app/shared/backend-services/reference-service/occupation-label.repository.ts [hunk 3]
```diff
@@ -50,6 +47,6 @@ export class OccupationLabelRepository {
       .set('prefix', query)
       .set('types', types.join(','))
       .set('resultSize', DEFAULT_RESPONSE_SIZE);
-    return this.http.get<OccupationLabelAutocomplete>(OCCUPATION_LABEL_RESOURCE_SEARCH_URL, { params });
+    return this.http.get<OccupationLabelAutocomplete>(OCCUPATION_LABEL_RESOURCE_SEARCH_URL, {params});
```
Comment (99999989) [inline alv-portal-ui/src/app/shared/backend-services/reference-service/occupation-label.repository.ts]: Same here: formatting is wrong afaik.

File: alv-portal-ui/src/app/shared/backend-services/reference-service/occupation-label.repository.ts (hunk 2 above)
Comment (ganqqwerty) [reply alv-portal-ui/src/app/shared/backend-services/reference-service/occupation-label.repository.ts]: weird. I did it again.

File: alv-portal-ui/src/app/shared/backend-services/reference-service/occupation-label.repository.ts (hunk 1 above)
Comment (ganqqwerty) [reply alv-portal-ui/src/app/shared/backend-services/reference-service/occupation-label.repository.ts]: agree

File: alv-portal-ui/src/app/shared/backend-services/reference-service/occupation-label.repository.ts [hunk 4]
```diff
@@ -23,21 +24,20 @@ export enum OccupationTypes {
...
   constructor(private http: HttpClient) {
   }
 
   getOccupationLabelsByKey(type: string, value: string, language: string): Observable<OccupationLabelData> {
     const cacheKey = `${type}_${value}_${language}`;
-    if (this.occupationLabelDataCache[cacheKey]) {
-      return of(this.occupationLabelDataCache[cacheKey]);
+    if (!this.occupationLabelDataCache[cacheKey]) {
+      // we cache the observable itself instead of the resolved value because the function is likely to be called in a loop
+      this.occupationLabelDataCache[cacheKey] = this.http.get<OccupationLabelData>(`${OCCUPATION_LABEL_RESOURCE_URL}/${type}/${value}`).pipe(
```
Comment (birom) [inline alv-portal-ui/src/app/shared/backend-services/reference-service/occupation-label.repository.ts]: What happens if there is a temporary error? Eg.: We restart the reference service?

Comment (99999989) [reply alv-portal-ui/src/app/shared/backend-services/reference-service/occupation-label.repository.ts]: I think the problem here is that we store the failed observable. So any subsequent subscription on this observable will fail, too. To force the subsequent subscriber to make a backend request, I catch the error and remove the observable from our cache.

Comment (ganqqwerty) [reply alv-portal-ui/src/app/shared/backend-services/reference-service/occupation-label.repository.ts]: I'm trying to test (with mocks at least) almost exactly the same solution, in theory it should work as Sven described.

**Pull request merged**